Source/JavaScriptCore/jit/AssemblyHelpers.cpp

```diff
@@ -608,43 +608,110 @@ void AssemblyHelpers::emitAllocateWithNonNullAllocator(GPRReg resultGPR, const J
     // - We *can* use RegisterSetBuilder::macroScratchRegisters on ARM.
 
     Jump popPath;
+    Jump zeroPath;
     Jump done;
-    
+
     if (allocator.isConstant())
         move(TrustedImmPtr(allocator.allocator().localAllocator()), allocatorGPR);
 
-    load32(Address(allocatorGPR, LocalAllocator::offsetOfFreeList() + FreeList::offsetOfRemaining()), resultGPR);
-    popPath = branchTest32(Zero, resultGPR);
+#if CPU(ARM) || CPU(ARM64)
+    auto dataTempRegister = getCachedDataTempRegisterIDAndInvalidate();
+#endif
+
+#if CPU(ARM64)
+    // On ARM64, we can leverage instructions like load-pair and shifted-add to make loading from the free list 
+    // and extracting interval information use less instructions.
+
+    // Bump allocation (fast path)
+    loadPairPtr(allocatorGPR, TrustedImm32(LocalAllocator::offsetOfFreeList() + FreeList::offsetOfIntervalStart()), resultGPR, scratchGPR);
```

Comment:
You should add a `static_assert` here that `FreeList::offsetOfIntervalEnd() == FreeList::offsetOfIntervalStart() + sizeof(char*)` because this `loadPairPtr` relies on it.

Reply:
I guess we can't `static_assert` this, but a normal `RELEASE_ASSERT` here shouldn't be a problem. I'll add one.